packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -20,6 +20,96 @@ typedef DiagnosticPropertiesTransformer = Iterable<DiagnosticsNode> Function(Ite
+///
+///  * [SubStackFilter], which uses this class to compare against [StackFrame]s.
+@immutable
+class PartialStackFrame {
+  /// Creates a new [PartialStackFrame] instance. All arguments are required and
+  /// must not be null.
+  const PartialStackFrame({
+    @required this.className,
+    @required this.method,
+    @required this.packagePath,
+  }) : assert(className != null),
+       assert(method != null),
+       assert(packagePath != null);
+
+  /// The class name for the method.
+  ///
+  /// On web, this is ignored, since class names are not available. Top level
+  /// methods should use the empty string.
+  final String className;
+
+  /// The method name for this frame line.
+  ///
+  /// On web, private methods are wrapped with `[]`.
+  final String method;
+
+  /// The relative path to the file for this method and class.
+  ///
+  /// This is required for disambiguation.
+  final String packagePath;
+
+  @override
+  String toString() {
+    if (kIsWeb) {
+      final String webMethod = method.startsWith('_') ? '[$method]' : method;
+      return '$packagePath: $webMethod';
+    }
+    return '$packagePath: $className.$method';
+  }
+}
+
+/// A class that matches sublists of stack frames for additional filtering
+/// of [FlutterError.defaultStackFilter].
+///
+/// This class provides a list of partial stack [frames] to match against and a
+/// replacement string for matches.
+///
+/// See also:
+///
+///   * [FlutterError.addDefaultStackFilter], a method to register additional
+///     stack filters for [FlutterError.defaultStackFilter].
+///   * [StackFrame], a class that can help with parsing stack frames.
+///   * [PartialStackFrame], a class that helps match partial method information
+///     to a stack frame.
+class SubStackFilter {
+  /// Creates a new SubStackFilter. All parameters are required and must not be
+  /// null.
+  const SubStackFilter({
+    @required this.frames,
+    @required this.replacement,
+  }) : assert(frames != null),
+       assert(replacement != null);
+
+  /// The shape of this repetative stack pattern.
+  final List<PartialStackFrame> frames;
+
+  /// The number of frames in this pattern.
+  int get numFrames => frames.length;
+
+  /// The string to replace the frames with.
+  ///
+  /// If the same replacement string is used multiple times in a row, the
+  /// [FlutterError.defaultStackFilter] will simply update a counter after this
+  /// line rather than repeating it.
+  final String replacement;
+
+  /// Checks whether the sublist of [StackFrame]s provided matches the [frames].
+  bool matchesFrames(List<StackFrame> stackFrames) {
+    if (stackFrames.length != numFrames) {
+      return false;
+    }
+    final String Function(StackFrame) mappingFunction = kIsWeb
+      ? (StackFrame frame) => '${frame.packagePath}: ${frame.method.startsWith('_') ? '[' + frame.method + ']' : frame.method}'
+      : (StackFrame frame) => '${frame.packagePath}: ${frame.className}.${frame.method}';
```

Comment:
can we implement this by passing the StackFrame to PartialStackFrame and asking if it matches? That way we'd avoid all the string manipulation.

Reply:
Done.